Engine/Utilities.cs

```diff
@@ -330,137 +342,169 @@ public Texture(int width, int height, IEnumerable<byte>? defaultDataPerPixel, Fo
                 Usage = ResourceUsage.Default,
                 CpuAccessFlags = CpuAccessFlags.None,
                 Format = textureFormat,
-                MipLevels = 1,
+                MipLevels = mipLevels,
                 OptionFlags = ResourceOptionFlags.Shared,
                 SampleDescription = new SampleDescription(1, 0)
             }, rectangles);
 
             if (defaultDataPerPixel != null)
                 Marshal.FreeHGlobal(dataPtr);
 
-            generateViews();
+            GenerateViews();
         }
-        private void generateViews()
+        private void GenerateViews()
         {
             Format format = texture.Description.Format;
             if (format == Format.R32_Typeless)
                 format = Format.R32_Float;
             BindFlags usage = texture.Description.BindFlags;
+
             int arraySize = texture.Description.ArraySize;
+            int mipLevels = texture.Description.MipLevels;
+
+            bool hasRenderTargets = usage.HasFlag(BindFlags.RenderTarget);
+            bool hasDepthStencils = usage.HasFlag(BindFlags.DepthStencil);
+            bool hasShaderResources = usage.HasFlag(BindFlags.ShaderResource);
+
+            if (hasRenderTargets) 
+                views.Add(CreateResourceViewCollection<RenderTargetView>(format, arraySize, mipLevels));
+            if (hasDepthStencils) 
+                views.Add(CreateResourceViewCollection<DepthStencilView>(format, arraySize, mipLevels));
+            if (hasShaderResources) 
+                views.Add(CreateResourceViewCollection<ShaderResourceView>(format, arraySize, mipLevels));
+        }
+
+        private ResourceViewCollection<T> CreateResourceViewCollection<T>(Format format, int arraySize, int mipLevels) 
+            where T : ResourceView
+        {
+            if(arraySize < 1)
+            {
+                throw new ArgumentNullException(nameof(arraySize));
+            }
 
-            if (usage.HasFlag(BindFlags.RenderTarget))
+            if(mipLevels < 1)
             {
-                if (arraySize > 1)
+                throw new ArgumentNullException(nameof(mipLevels));
```

Comment:
argument is not null, argument is out of range, throw ArgumentOutOfRangeException

Reply:
Fix